src/test/validation_chainstatemanager_tests.cpp

```diff
@@ -101,4 +101,58 @@ BOOST_AUTO_TEST_CASE(chainstatemanager)
     WITH_LOCK(::cs_main, manager.Unload());
 }
 
+//! Test rebalancing the caches associated with each chainstate.
+BOOST_AUTO_TEST_CASE(chainstatemanager_rebalance_caches)
+{
+    ChainstateManager manager;
+    int max_cache = 10000;
+    manager.m_total_coinsdb_cache = max_cache;
+    manager.m_total_coinstip_cache = max_cache;
+
+    std::vector<CChainState*> chainstates;
+
+    // Create a legacy (IBD) chainstate.
+    //
+    ENTER_CRITICAL_SECTION(cs_main);
+    CChainState& c1 = manager.InitializeChainstate();
+    LEAVE_CRITICAL_SECTION(cs_main);
```

Comment:
In commit "add ChainstateManager::MaybeRebalanceCaches()" (db5e36a033e642599db808f59d16d03a40860f17)

This is interesting. Once we update to c++14 or 17 we should be able to use WITH_LOCK like

```c++
CChainState& c1 = WITH_LOCK(cs_main, return manager.InitializeChainstate());
```

I think it changing it wouldn't work now without decltype(auto)

Reply:
What is different between this potential WITH_LOCK use, and the existing ones that return a value?

Reply:
I don't know, but maybe C++11 lambda can not return references? If yes, this can be solved with


```cpp
CChainState& c1 = *WITH_LOCK(cs_main, return &manager.InitializeChainstate());

Reply:
Yes, just what marco said, auto return type won't deduce a reference, and I think the pointer trick would work here, so it might nice to incorporate if the PR is gets updated again